debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/jsqlparser/SimpleDmlParser.java

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.oracle.jsqlparser;
+
+import java.io.StringReader;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.connector.oracle.antlr.listener.ParserUtils;
+import io.debezium.connector.oracle.logminer.OracleChangeRecordValueConverter;
+import io.debezium.connector.oracle.logminer.valueholder.LogMinerColumnValue;
+import io.debezium.connector.oracle.logminer.valueholder.LogMinerColumnValueImpl;
+import io.debezium.connector.oracle.logminer.valueholder.LogMinerColumnValueWrapper;
+import io.debezium.connector.oracle.logminer.valueholder.LogMinerDmlEntry;
+import io.debezium.connector.oracle.logminer.valueholder.LogMinerDmlEntryImpl;
+import io.debezium.data.Envelope;
+import io.debezium.relational.Column;
+import io.debezium.relational.Table;
+import io.debezium.relational.TableId;
+import io.debezium.relational.Tables;
+import io.debezium.text.ParsingException;
+
+import net.sf.jsqlparser.JSQLParserException;
+import net.sf.jsqlparser.expression.Alias;
+import net.sf.jsqlparser.expression.Expression;
+import net.sf.jsqlparser.expression.ExpressionVisitorAdapter;
+import net.sf.jsqlparser.expression.operators.relational.EqualsTo;
+import net.sf.jsqlparser.expression.operators.relational.ExpressionList;
+import net.sf.jsqlparser.expression.operators.relational.IsNullExpression;
+import net.sf.jsqlparser.expression.operators.relational.ItemsList;
+import net.sf.jsqlparser.expression.operators.relational.ItemsListVisitorAdapter;
+import net.sf.jsqlparser.parser.CCJSqlParserManager;
+import net.sf.jsqlparser.statement.Statement;
+import net.sf.jsqlparser.statement.delete.Delete;
+import net.sf.jsqlparser.statement.insert.Insert;
+import net.sf.jsqlparser.statement.update.Update;
+
+/**
+ * This class does parsing of simple DML: insert, update, delete.
+ * Log Miner supplies very simple syntax , that this parser should be sufficient to parse those.
+ * It does no support joins, merge, sub-selects and other complicated cases, which should be OK for Log Miner case
+ */
```

Comment:
Completely agree and this is why I use simple RegEx statements to get rid of jsqlparser. jsqlparser is very slow... But maybe you are using the fast parts...
For example `Regex("(?i)^UPDATE .* SET (.*) WHERE (.*)$")` and then split on AND

Reply:
I understand its likely slow and the contributor has another implementation coming but we needed to get this rebased & integrated.  Once we have the parser code submitted, I'd be interested in hearing from you on performance if you have the bandwidth.

Reply:
I did not do a thorough and detailed performance test but when I used jsqlParser the CPU went to 100% all the time and parsing was slow (I reached about 400 msg/sec). Then I changed to RegEx and now the CPU stays relatively low and am at about 4000 msg/sec.

Reply:
Having a cheap parser would be great.

Reply:
First thing I do is remove the newlines and replace them with a placeholder for later restoring.
```
private val newlineInFieldPattern = Regex("('[^'\n]*)\r?\n(?!(([^']*'){2})*[^']*\$)")
val cleanedSqlRedo = sqlRedo.replace(newlineInFieldPattern, "$1$NEWLINE_STRING").replace("\n", " ")
```
Then for the different types of operations different regexes:
```
val regex = Regex("(?i)^INSERT INTO [^(]*\\((.*)\\) VALUES \\((.*)\\)$")
val setAndWhereRegEx = Regex("(?i)^UPDATE .* SET (.*) WHERE (.*)$")
val regex = Regex("(?i)^DELETE FROM .* WHERE (.*)$")
```

This separates the before and after values and groups them. Some simple splitting on 'AND' and fixing words like TIMESTAMP, IS NULL, etc

```
private val quotePattern = Regex("^[\\s]*['\"](.*)['\"][\\s]*$")
private val newlineInFieldPattern = Regex("('[^'\n]*)\r?\n(?!(([^']*'){2})*[^']*\$)")
val cleaned = str
                    .replace(timestampPattern, "") // remove TIMESTAMP word
                    .replace(quotePattern, "$1") // remove starting and trailing ' and "
                    .replace(NEWLINE_STRING, "\n")
                    .trim() // remove any surrounding spaces
```

Reply:
```
@Service
class ResultSetService(
        @Value("\${app.source.brand}") val brand: String,
        @Value("\${app.source.domain}") val domain: String,
        @Value("\${app.source.datastore}") val datastore: String,
        @Value("\${app.source.timezone}") val timezone: String
) {
    companion object {
        private const val NULL_STRING = "NULL"
        private const val NEWLINE_STRING = "#NL#"

        // Match 'comma' but only if not enclosed in quotes
        val csvPattern = Regex(",(?=(?:[^']*'[^']*')*[^']*$)")

        private val timestampPattern = Regex("^[\\s]*TIMESTAMP[\\s]+")
        private val quotePattern = Regex("^[\\s]*['\"](.*)['\"][\\s]*$")
        private val newlineInFieldPattern = Regex("('[^'\n]*)\r?\n(?!(([^']*'){2})*[^']*\$)")

        fun cleanString(str: String): String {
            val timestampFound = timestampPattern.containsMatchIn(str)

            val cleaned = str
                    .replace(timestampPattern, "") // remove TIMESTAMP word
                    .replace(quotePattern, "$1") // remove starting and trailing ' and "
                    .replace(NEWLINE_STRING, "\n")
                    .trim() // remove any surrounding spaces

            // If it is a TIMESTAMP replace it with the number of milliseconds since epoch UTC
            return if (timestampFound)
                Timestamp.valueOf(cleaned).toInstant().toEpochMilli().toString()
            else
                cleaned
        }
    }

    fun getSource(dataSet: String): Source {
        return Source(
                brand = brand,
                domain = domain,
                datastore = datastore,
                dataset = dataSet,
                schema = "$dataSet-v1 schema",
                timezone = timezone
        )
    }

    fun mapResultToData(operation: SqlOperation, sqlRedo: String): TransactionEntry {
        val cleanedSqlRedo = sqlRedo.replace(newlineInFieldPattern, "$1$NEWLINE_STRING").replace("\n", " ")

        val (beforeValues, afterValues) = when (operation) {
            SqlOperation.INSERT -> handleInsert(cleanedSqlRedo)
            SqlOperation.UPDATE -> handleUpdate(cleanedSqlRedo)
            SqlOperation.DELETE -> handleDelete(cleanedSqlRedo)
            SqlOperation.SELECT_FOR_UPDATE -> error("$operation not implemented yet!")
            SqlOperation.COMMIT -> error("$operation should not be reachable! Fix the code!")
            SqlOperation.ROLLBACK -> error("$operation should not be reachable! Fix the code!")
            SqlOperation.START -> error("$operation should not be reachable! Fix the code!")
        }

        val fields = beforeValues.keys.union(afterValues.keys).toTypedArray()
        val beforeList = fields.map { field -> beforeValues[field] }.toTypedArray()
        val afterList = fields.map {field ->
            val beforeValue = beforeValues[field]
            val afterValue = afterValues[field]
            if(afterValue == beforeValue) beforeValue else afterValue
        }.toTypedArray()

        return TransactionEntry(
                operation = operation,
                fields = fields,
                before = beforeList,
                after = afterList
        )
    }

    private fun handleInsert(sqlRedo: String): Pair<Map<String, String>, Map<String, String>> {
        // Always in the form 'INSERT INTO "<schema>"."<table>"("<field_1>", ..., "<field_n>") VALUES (<value_1>, ..., <value_n>)'
        val regex = Regex("(?i)^INSERT INTO [^(]*\\((.*)\\) VALUES \\((.*)\\)$")
        val matches = regex.find(sqlRedo)!!
        val fieldsGroup = matches.groups[1]!!.value
        val valuesGroup = matches.groups[2]!!.value

        // Get the fields and the values
        val fields = fieldsGroup
                .split(csvPattern)
                .map { fieldName -> cleanString(fieldName) }
        val values = valuesGroup
                .split(csvPattern)
                .map { value -> cleanString(value) }

        // Create the map with field -> value pairs
        val afterValues = fields
                .mapIndexed { index, field -> field to values[index] }
                .toMap()

        val beforeValues = mapOf<String, String>()
        return Pair(beforeValues, afterValues)
    }

    private fun handleUpdate(sqlRedo: String): Pair<Map<String, String>, Map<String, String>> {
        // Almost always in the form:
        //   UPDATE "<schema>"."<table>"
        //   SET "<field_1>" = "<value_1>", ..., "<field_n>" = "<value_n>"
        //   WHERE "<field_1>" = "<value_1>" AND ... AND "<field_n>" = "<value_n>"
        val setAndWhereRegEx = Regex("(?i)^UPDATE .* SET (.*) WHERE (.*)$")
        val matchesSetAndWhere = setAndWhereRegEx.find(sqlRedo)

        var matchesSetWithoutWhere: MatchResult? = null
        if (matchesSetAndWhere == null) {
            // Or else might be without the where:
            //   UPDATE "<schema>"."<table>"
            //   SET "<field_1>" = "<value_1>", ..., "<field_n>" = "<value_n>"
            val setWithoutWhereRegEx = Regex("(?i)^UPDATE .* SET (.*)$")
            matchesSetWithoutWhere = setWithoutWhereRegEx.find(sqlRedo) ?: error("handleUpdate: unable to handle $sqlRedo")
        }
        val setStatementsGroup = matchesSetAndWhere?.groups?.get(1)?.value ?: matchesSetWithoutWhere!!.groups[1]!!.value
        val whereStatementGroup = matchesSetAndWhere?.groups?.get(2)?.value ?: ""

        // Extract the important parts of the statement
        val setStatements = setStatementsGroup
                .split(csvPattern)
                .map { pair -> pair.split(" = ").let { Pair(cleanString(it[0]), cleanString(it[1])) } }
        val whereStatements = matchesSetAndWhere?.let {
            whereStatementGroup
                    .split(" AND ", ignoreCase = true)
                    .map { pair -> pair.split(" = ", " IS ").let { Pair(cleanString(it[0]), cleanString(it[1])) } }
        } ?: emptyList()

        // Before values are just the WHERE values. Logminer sends all field in the WHERE
        val beforeValues = whereStatements.toMap().toMutableMap()

        // Use the beforeValues as start for the afterValues and overwrite with changes
        val afterValues = beforeValues.toMutableMap()
        setStatements.forEach { (field, value) -> afterValues[field] = value }

        // If both before and after field contains 'null' then just ignore the field
        beforeValues
                .filter { (field, value) -> value == NULL_STRING && (afterValues[field]?.let { it == NULL_STRING } ?: true) }
                .forEach { (field, _) ->
                    beforeValues.remove(field)
                    afterValues.remove(field)
                }

        return Pair(beforeValues, afterValues)
    }

    private fun handleDelete(sqlRedo: String): Pair<Map<String, String>, Map<String, String>> {
        // Always in the form 'DELETE FROM "<schema>"."<table>" WHERE "<field_1>" = "<value_1>" AND ... AND "<field_n>" = "<value_n>"'
        val regex = Regex("(?i)^DELETE FROM .* WHERE (.*)$")
        val matches = regex.find(sqlRedo)!!
        val whereStatementGroup = matches.groups[1]!!.value

        val whereStatements = whereStatementGroup
                .split(" AND ", ignoreCase = true)
                .map { fieldValuePair -> fieldValuePair.split(" = ", " IS ").let { Pair(cleanString(it[0]), cleanString(it[1])) } }

        // Before values are just the WHERE values. Logminer sends all field in the WHERE
        val beforeValues = whereStatements.toMap()

        val afterValues = mapOf<String, String>()
        return Pair(beforeValues, afterValues)
    }
}
```

Reply:
I rewrote the above code to get rid of the RegEx parts. They were introducing a lot of garbage and were not fast enough. I now do a sqlRedo.forEach { c -> ... } and so parse the string only once. This is ten times as fast. And even fixes the problem of having ' WHERE ' in your values that failed with the RegEx.
If wanted I can share the code... As inspiration. Still needs to be thoroughly tested though.

Reply:
Thanks a lot, @MilovdZee, that's great input!

@Naros, could you capture this in a follow-up Jira issue? Indeed we might get away with just a regexp if the DML structure is so uniform and simple.

Reply:
For me the RegEx was not fast enough. And keep in minds that it was also not fail safe. But it was about the concept. For example my RegEx failed if there was a " WHERE " between quotes in one of the values.
I now do it even more manual and am able to handle up to 25000 UPDATES a second on my laptop. For INSERT this is even close to a 100.000. The only big problem is that the code is quite unreadable...

```
@Service
class ResultSetService(
        @Value("\${app.source.brand}") val brand: String,
        @Value("\${app.source.domain}") val domain: String,
        @Value("\${app.source.datastore}") val datastore: String,
        @Value("\${app.source.timezone}") val timezone: String
) {
    companion object {
        private const val MAX_STRING_SIZE = 10000
        private const val NULL_STRING = "NULL"
        private const val NOT_IN_QUOTES = '-'

        private val setArray = "SET".toCharArray()
        private val whereArray = "WHERE".toCharArray()
    }

    // Buffer used while parsing the sql statements
    // The result of using a shared buffer is that the service is not thread-safe...
    private val string = CharArray(MAX_STRING_SIZE)
    private var stringIndex = 0

    fun getSource(dataSet: String): Source {
        return Source(
                brand = brand,
                domain = domain,
                datastore = datastore,
                dataset = dataSet,
                schema = "$dataSet-v1 schema",
                timezone = timezone
        )
    }

    fun mapResultToData(operation: SqlOperation, sqlRedo: String): TransactionEntry {
        val (beforeValues, afterValues) = when (operation) {
            SqlOperation.INSERT -> handleInsert(sqlRedo)
            SqlOperation.UPDATE -> handleUpdate(sqlRedo)
            SqlOperation.DELETE -> handleDelete(sqlRedo)
            SqlOperation.SELECT_FOR_UPDATE -> error("$operation not implemented yet!")
            SqlOperation.COMMIT -> error("$operation should not be reachable! Fix the code!")
            SqlOperation.ROLLBACK -> error("$operation should not be reachable! Fix the code!")
            SqlOperation.START -> error("$operation should not be reachable! Fix the code!")
        }

        return TransactionEntry(
                operation = operation,
                before = beforeValues,
                after = afterValues
        )
    }

    private fun handleInsert(sqlRedo: String): Pair<Map<String, String>, Map<String, String>> {
        // Always in the form 'INSERT INTO "<schema>"."<table>"("<field_1>", ..., "<field_n>") VALUES (<value_1>, ..., <value_n>)'
        // So:
        // - two blocks between braces. First block contains the fields and the second block contains the values
        // - Steps:
        //   - find the first opening brace
        //   - read comma separated fields until the closing brace
        //   - find next opening brace
        //   - read comma separated values until the closing brace
        //   - combine fields and values into result

        var inQuotes = false
        var inBrackets = false
        var inBlock = false
        var timestamp = false
        var state = InsertBlockSequence.FIELDS

        val fields = mutableListOf<String>()
        val values = mutableListOf<String>()

        sqlRedo.forEach { c ->
            if (c == '"' || c == '\'') {
                // Remove quotes from the string and keep track of if we are inside quotes
                inQuotes = !inQuotes
            } else if (!inQuotes && c == '(') {
                // remove the brackets from the string keep track of if we are inside brackets
                inBrackets = true
            } else if (inBrackets) {
                if (inBlock && !inQuotes) {
                    if (c == ',' || c == ')' || c == ' ') {
                        // Handle end of block
                        inBlock = false
                        when (state) {
                            InsertBlockSequence.FIELDS -> {
                                fields.add(String(string, 0, stringIndex))
                                stringIndex = 0
                            }
                            InsertBlockSequence.VALUES -> {
                                val value = if (timestamp)
                                    Timestamp.valueOf(String(string, 0, stringIndex)).toInstant().toEpochMilli().toString()
                                else
                                    String(string, 0, stringIndex)
                                timestamp = value == "TIMESTAMP"
                                if (!timestamp) {
                                    values.add(value)
                                }
                                stringIndex = 0
                            }
                        }

                        // Closing bracket, so go to the next state
                        if (c == ')') {
                            state = InsertBlockSequence.VALUES
                            inBrackets = false
                        }
                    }
                } else if (c != ' ' && c != '\n') {
                    // It is not a space so must be inside some kind of block
                    inBlock = true
                }

                // Everything within quotes is always part of the block
                if (inBlock || inQuotes) {
                    string[stringIndex++] = c
                    if(stringIndex == MAX_STRING_SIZE) error("handleInsert: string to long to fit in buffer ($sqlRedo)")
                }
            }
        }

        val beforeValues = mapOf<String, String>()
        val afterValues = fields.mapIndexed { index, field -> field to values[index] }.toMap()
        return Pair(beforeValues, afterValues)
    }

    private fun handleUpdate(sqlRedo: String): Pair<Map<String, String>, Map<String, String>> {
        // Almost always in the form:
        //   UPDATE "<schema>"."<table>"
        //   SET "<field_1>" = '<value_1>', ..., "<field_n>" = '<value_n>'
        //   WHERE "<field_1>" = '<value_1>' AND ... AND "<field_n>" = '<value_n>'
        //
        // Notes:
        // - 'value' could be "NULL" without quotes
        // - 'value' could be "TIMESTAMP ' 2020-04-23 00:00:00'"
        // - "=" could also be "IS"

        val (beforeMap, afterMap) = extractBeforeAndAfterMap(sqlRedo, false)

        // Add missing fields to the afterMap
        beforeMap.entries
                .filter { (_, value) -> value != NULL_STRING }
                .filter { (field, _) -> !afterMap.containsKey(field) }
                .forEach { (field, value) -> afterMap[field] = value }

        // If both before and after field contains 'null' then just ignore the field
        beforeMap
                .filter { (field, value) -> value == NULL_STRING && (afterMap[field]?.let { it == NULL_STRING } ?: true) }
                .forEach { (field, _) ->
                    beforeMap.remove(field)
                    afterMap.remove(field)
                }

        return Pair(beforeMap, afterMap)
    }

    private fun handleDelete(sqlRedo: String): Pair<Map<String, String>, Map<String, String>> {
        // Always in the form 'DELETE FROM "<schema>"."<table>" WHERE "<field_1>" = "<value_1>" AND ... AND "<field_n>" = "<value_n>"'
        return extractBeforeAndAfterMap(sqlRedo, true)
    }

    private fun extractBeforeAndAfterMap(sqlRedo: String, skipSet: Boolean): Pair<MutableMap<String, String>, MutableMap<String, String>> {
        var setIndex = 0
        var whereIndex = 0
        var setSeen = false
        var whereSeen = false

        val afterMap = mutableMapOf<String, String>()
        val beforeMap = mutableMapOf<String, String>()

        var timestamp = false
        var inQuotesChar = NOT_IN_QUOTES
        var key = ""
        var state = UpdateBlockSequence.KEY // Start by looking for a key
        var inBlock = false
        val sqlRedoLength = sqlRedo.length
        for (charIndex in 0..sqlRedoLength) {
            val c = if (charIndex == sqlRedoLength) '\n' else sqlRedo[charIndex]

            // Search for SET
            if (!setSeen && !skipSet) {
                if (setArray[setIndex] == c) {
                    setIndex++
                    if (setIndex == setArray.size) {
                        setSeen = true
                    }
                } else setIndex = 0
            } else {
                if (c == '"' || c == '\'') {
                    // Keep track of if we are inside quotes
                    if (inQuotesChar == NOT_IN_QUOTES) {
                        // Start of quoted part
                        inQuotesChar = c
                    } else if (inQuotesChar == c) {
                        // End of quoted part
                        inQuotesChar = NOT_IN_QUOTES
                    }
                }

                // Search for WHERE
                if (!whereSeen) {
                    if (inQuotesChar == NOT_IN_QUOTES && whereArray[whereIndex] == c) {
                        whereIndex++
                        if (whereIndex == whereArray.size) {
                            whereSeen = true
                            continue // handle the next character
                        }
                    } else whereIndex = 0
                }

                if (whereSeen || !skipSet) {
                    if (inBlock && inQuotesChar == NOT_IN_QUOTES) {
                        // Handle end of block
                        if (c == ',' || c == '=' || c == ' ' || c == '\n' || c == '"' || c == '\'') {
                            inBlock = false
                            when (state) {
                                UpdateBlockSequence.KEY -> {
                                    key = String(string, 0, stringIndex)
                                    stringIndex = 0
                                    state = if (c == ' ' || c == '\n' || c == '"' || c == '\'') UpdateBlockSequence.SEPERATOR_AFTER_KEY else UpdateBlockSequence.VALUE
                                }
                                UpdateBlockSequence.SEPERATOR_AFTER_KEY -> {
                                    state = UpdateBlockSequence.VALUE
                                }
                                UpdateBlockSequence.VALUE -> {
                                    val value = if (timestamp)
                                        Timestamp.valueOf(String(string, 0, stringIndex)).toInstant().toEpochMilli().toString()
                                    else
                                        String(string, 0, stringIndex)
                                    timestamp = value == "TIMESTAMP"
                                    if (!timestamp) {
                                        if (!whereSeen) afterMap[key] = value else beforeMap[key] = value
                                        state = if (c == ' ' || c == '\n' || c == '"' || c == '\'') UpdateBlockSequence.SEPARATOR_AFTER_VALUE else UpdateBlockSequence.KEY
                                    }
                                    stringIndex = 0
                                }
                                UpdateBlockSequence.SEPARATOR_AFTER_VALUE -> {
                                    state = UpdateBlockSequence.KEY
                                }
                            }
                        }
                    } else if (inQuotesChar != NOT_IN_QUOTES || (c != ' ' && c != '\n')) {
                        // It is not a space so must be inside some kind of block.
                        // And everything between quotes is part of a block
                        inBlock = true
                    }

                    if ((state == UpdateBlockSequence.KEY || state == UpdateBlockSequence.VALUE) && inBlock && c != '"' && c != '\'') {
                        string[stringIndex++] = c
                        if(stringIndex == MAX_STRING_SIZE) error("extractBeforeAndAfterMap: string to long to fit in buffer ($sqlRedo)")
                    }
                }
            }
        }
        return Pair(beforeMap, afterMap)
    }
}

private enum class InsertBlockSequence {
    FIELDS,
    VALUES
}

private enum class UpdateBlockSequence {
    KEY,
    SEPERATOR_AFTER_KEY,
    VALUE,
    SEPARATOR_AFTER_VALUE
}


`

Reply:
I've included this as part of [DBZ-2532](https://issues.redhat.com/browse/DBZ-2532).  Please feel free to leave any additional comments there as I've referenced the above code snippet for reference to this discussion.